diff --git a/nixos/modules/system/boot/loader/generic-extlinux-compatible/extlinux-conf-builder.sh b/nixos/modules/system/boot/loader/generic-extlinux-compatible/extlinux-conf-builder.sh
index 1a0da005029..7a82bbeeb19 100644
--- a/nixos/modules/system/boot/loader/generic-extlinux-compatible/extlinux-conf-builder.sh
+++ b/nixos/modules/system/boot/loader/generic-extlinux-compatible/extlinux-conf-builder.sh
@@ -128,6 +128,8 @@ DEFAULT nixos-default
 
 MENU TITLE ------------------------------------------------------------
 TIMEOUT $timeout
+# TODO: Check whether this is an issue with uboot
+UI menu.c32
 EOF
 
 addEntry $default default >> $tmpFile
